In [60]:
from astropy import units as u
from astroquery.atomic import AtomicLineList
import astroquery.atomic as a
from astropy.table import conf

# Possible IDs for the deep neutral lines

These are the lines, whose spatial distribution is similar to C I 8727 and the H_2 2.12 micron line. 

## Lines around the neutral 8152 line

This is the shortest wavelength of these lines, which we are lookin at first.  We are also interested in other lines in the same wavelength range, even if they are form the ionized gas. 

There are certainly some He lines there.

In [2]:
wavelength_range = (8152 * u.Angstrom, 70 * u.Angstrom)
res = AtomicLineList.query_object(
    wavelength_range=wavelength_range,
    wavelength_type='Air',
    minimal_abundance="6",
    depl_factor="0",
    element_spectrum="\n".join(['H', 'He']),
    #get_query_payload=True,
)

/Users/will/miniconda3/envs/py39/lib/python3.9/site-packages/astroquery/atomic/core.py:309: GuessedAtParserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system ("html5lib"). This usually isn't a problem, but if you run this code on another system, or in a different virtual environment, it may use a different parser and behave differently.

The code that caused this warning is on line 309 of the file /Users/will/miniconda3/envs/py39/lib/python3.9/site-packages/astroquery/atomic/core.py. To get rid of this warning, pass the additional argument 'features="html5lib"' to the BeautifulSoup constructor.

  bs = BeautifulSoup(response.text, 'html5')
/Users/will/miniconda3/envs/py39/lib/python3.9/site-packages/astroquery/atomic/core.py:249: GuessedAtParserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system ("html5lib"). This usually isn't a problem, but if you run this code on another system

In [3]:
res

<Table length=19>
LAMBDA AIR ANG SPECTRUM  TT  ... J J    A_ki      LEVEL ENERGY  CM 1  
   float64       str4   str2 ... str3 float64           str23         
-------------- -------- ---- ... ---- -------- -----------------------
       8093.08     He I   E1 ...  2-1  29100.0 185564.67 -   197917.50
       8093.09     He I   E1 ...  1-1  17500.0 185564.69 -   197917.50
       8093.27     He I   E1 ...  0-1   5830.0 185564.96 -   197917.50
       8094.08     He I   E1 ...  0-1 144000.0 184864.93 -   197216.24
       8127.09     He I   E1 ...  2-1  35100.0 185564.67 -   197865.81
       8127.11     He I   E1 ...  1-1  21100.0 185564.69 -   197865.81
       8127.29     He I   E1 ...  0-1   7030.0 185564.96 -   197865.81
       8155.59     He I   E1 ...  2-*  68500.0 185564.67 -   197822.82
       8155.61     He I   E1 ...  1-*  41100.0 185564.69 -   197822.82
       8155.79     He I   E1 ...  0-*  13700.0 185564.96 -   197822.82
       8168.64     He I   E1 ...  2-1  42900.0 185564.67 -   197803.23
       8168.66     He I   E1 ...  1-1  25700.0 185564.69 -   197803.23
       8168.84     He I   E1 ...  0-1   8580.0 185564.96 -   197803.23
       8203.85     He I   E1 ...  2-*  84300.0 185564.67 -   197750.71
       8203.86     He I   E1 ...  1-*  50600.0 185564.69 -   197750.71
       8204.04     He I   E1 ...  0-*  16900.0 185564.96 -   197750.71
       8220.22     He I   E1 ...  2-1  53100.0 185564.67 -   197726.45
       8220.23     He I   E1 ...  1-1  31900.0 185564.69 -   197726.45
       8220.42     He I   E1 ...  0-1  10600.0 185564.96 -   197726.45

In [66]:
wavelength_range = (8152 * u.Angstrom, 70 * u.Angstrom)
ionlist = " I-IV\n".join(
    [
        "C", 
        "N", 
        "O", 
        "S",
        "Si",
        "Cl",
        "Fl",
        "Fe",
        "Ne",
        "Ar",
        "Ni",
        "Ca",
        "Mg",
    ]
)
res = AtomicLineList.query_object(
    wavelength_range=wavelength_range,
    wavelength_type='Air',
    lower_level_energy_range=3 * u.eV,
    #transitions=aa.Transition.nebular,
    element_spectrum=ionlist,
    #get_query_payload=True,
)

In [67]:
res

<Table length=15>
LAMBDA AIR ANG SPECTRUM  TT  ...   J J    A_ki   LEVEL ENERGY  CM 1  
   float64       str7   str2 ...   str8   str4         str22         
-------------- -------- ---- ... -------- ---- ----------------------
      8105.651    Ni I]   E1 ...      2-1 None 15609.84 -    27943.52
     8108.3199    Fe I]   E1 ...      4-3 None 21999.13 -    34328.75
      8111.975   [Ni I]   M1 ...      2-1 None  3409.94 -    15734.00
     8112.1609    Fe I]   E1 ...      5-4 None 21715.73 -    34039.51
      8119.142  [Fe II]   M1 ...  5/2-3/2 None 13474.41 -    25787.60
     8129.3131    Fe I]   E1 ...      3-2 None 22249.43 -    34547.21
     8151.3424   [Fe I]   M1 ...      2-2 None   704.01 -    12968.55
     8180.0809    Fe I]   E1 ...      3-4 None 20874.48 -    33095.94
      8184.503  [Fe II]   E2 ... 13/2-9/2 None 21251.61 -    33466.46
      8188.077    Ni I]   E1 ...      1-1 None 15734.00 -    27943.52
      8194.531   [Ni I]   M1 ...      2-2 None  3409.94 -    15609.84
       8199.03  [Fe II]   E2 ...  3/2-7/2 None 21308.04 -    33501.25
       8201.74   [Ni I]   M1 ...      3-2 None  1332.16 -    13521.35
     8204.1023    Fe I]   E1 ...      4-4 None  7376.76 -    19562.44
     8204.9366    Fe I]   E1 ...      3-2 None  7728.06 -    19912.49

In [68]:
wavelength_range = (8152 * u.Angstrom, 70 * u.Angstrom)
ionlist = "\n".join([
    "N I", "N II", 
    "C I", "C II", 
    "O I", "O II", 
    "Si I", "Si II", 
    "Ca I", "Ca II", 
])

res = AtomicLineList.query_object(
    wavelength_range=wavelength_range,
    wavelength_type='Air',
    #lower_level_energy_range=3 * u.eV,
    #transitions=aa.Transition.nebular,
    element_spectrum=ionlist,
    #get_query_payload=True,
)
res.show_in_notebook()

## Trying to find IDs for the lines

First look for all the neutral lines of common elements that fall in the wavelength range where we see the lines.


### Forbidden neutral lines of all the common elements

Start off with just the forbidden lines, so we do not get too many. Also, impose an upper energy level cutoff of less than 4 eV, which would make sense if they wer excited collisionally.

In [99]:
wavelength_range = (
    8100 * u.Angstrom, 9200 * u.Angstrom
)
elements = "\n".join(
    [
        "C", "N", "O",
        "Mg", "Ca", "Na", 
        "S", "Si", "P",
        "F", "Cl",
        "Ar", "Xe",
        "Fe", "Ni", "Co",
    ]
)
res = AtomicLineList.query_object(
    wavelength_range=wavelength_range,
    wavelength_type='Air',
    upper_level_energy_range=10 * u.eV,
    element_spectrum=elements,
    transitions=aa.Transition.nebular,
)

/Users/will/miniconda3/envs/py39/lib/python3.9/site-packages/astroquery/atomic/core.py:249: GuessedAtParserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system ("html5lib"). This usually isn't a problem, but if you run this code on another system, or in a different virtual environment, it may use a different parser and behave differently.

The code that caused this warning is on line 249 of the file /Users/will/miniconda3/envs/py39/lib/python3.9/site-packages/astroquery/atomic/core.py. To get rid of this warning, pass the additional argument 'features="html5lib"' to the BeautifulSoup constructor.

  data = StringIO(BeautifulSoup(response.text, 'html5').find('pre').text.strip())


Add a new column that gives the ion stage: I, II, III, IV, etc

In [100]:
res["stage"] = [s.strip("[]").split()[-1] for s in res["SPECTRUM"]]

And use it to select just the neutral ions, since the API does not allow this to be done on the query. 

In [101]:
res[res["stage"]=="I"].show_in_notebook()

#### Confident ID: [Fe I] 8151.3424

I am pretty sure of this one. 

#### Other potential matches

So there are some partial overlaps with the observed wavelengths here. 

For instance, 9020 might be [Co I] 9019.65

But most of the wavelengths are misses.



In [106]:
(1 * u.eV).to(1/u.cm, equivalencies=u.equivalencies.spectral())

<Quantity 8065.54393735 1 / cm>

### Permitted lines of selected neutral atoms

Next thing to try is to expand search to include permitted lines as well as forbidden ones. If they are not from autoionizing levels, then they should have upper level energies less than the ionization potential of the neutral atom, which for these elements is about 6 eV.

I have put a limit of 4 eV to start with, so I do not get too many hits. 

The excitation could be collisions, fluorescence, or recombination.  In the case of collisions, the upper level cannot be very high since w have a very low T, probably hundreds of K rather than thousands. 

But for the C I line, recombinations are thought to dominate. If the same is true of the other lines, then the brightness will be proportional to the abundance, more or less I should think. 

The C I 8727 line is not a resonance line, and is 10 times weaker than C I 9850 (which is), according to Cesarsky 1982A&A...113L...7C

This means there is room for observable recombination lines from less abundant elements, but they will be most likely resonant lines since those are the ones that capture the greatest fraction of the total recombination rate. 

In [111]:
res = AtomicLineList.query_object(
    wavelength_range=wavelength_range,
    wavelength_type='Air',
    upper_level_energy_range=4 * u.eV,
    element_spectrum="Fe",
    #transitions=aa.Transition.nebular,
)
res["stage"] = [s.strip("[]").split()[-1] for s in res["SPECTRUM"]]

/Users/will/miniconda3/envs/py39/lib/python3.9/site-packages/astroquery/atomic/core.py:249: GuessedAtParserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system ("html5lib"). This usually isn't a problem, but if you run this code on another system, or in a different virtual environment, it may use a different parser and behave differently.

The code that caused this warning is on line 249 of the file /Users/will/miniconda3/envs/py39/lib/python3.9/site-packages/astroquery/atomic/core.py. To get rid of this warning, pass the additional argument 'features="html5lib"' to the BeautifulSoup constructor.

  data = StringIO(BeautifulSoup(response.text, 'html5').find('pre').text.strip())


In [112]:
res[res["stage"]=="I"].show_in_notebook()

So that is with an upper limit of 4 eV for the energy, which gives 39 candidate lines of Fe I. If I use 7 eV instead, then I get 10 times as many lines. But then, the question is: why do we see only some of those (perhaps they vary greatly in strength, but the database does not have A values for the majority).

This gives a potential match for the 9146 line of Fe I 9146.1285, but not for anything else. 

We can try it again for the other likely elements.

Try looking first at resonance lines of neutral atoms. I am restricting the lower energy level to less than 0.1 eV, so that will include the fine structure splitting of the ground configuration. 


In [118]:
res = AtomicLineList.query_object(
    wavelength_range=wavelength_range,
    wavelength_type='Air',
    lower_level_energy_range=0.1 * u.eV,
    #element_spectrum="Fe",
    #transitions=aa.Transition.nebular,
)
res["stage"] = [s.strip("[]").split()[-1] for s in res["SPECTRUM"]]

/Users/will/miniconda3/envs/py39/lib/python3.9/site-packages/astroquery/atomic/core.py:249: GuessedAtParserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system ("html5lib"). This usually isn't a problem, but if you run this code on another system, or in a different virtual environment, it may use a different parser and behave differently.

The code that caused this warning is on line 249 of the file /Users/will/miniconda3/envs/py39/lib/python3.9/site-packages/astroquery/atomic/core.py. To get rid of this warning, pass the additional argument 'features="html5lib"' to the BeautifulSoup constructor.

  data = StringIO(BeautifulSoup(response.text, 'html5').find('pre').text.strip())


In [119]:
res[res["stage"]=="I"].show_in_notebook()

So that gives no good matches, apart from the Fe I line that we already knew about. 

Just to make sure, we eill look at singly ionized species too, just in case there is one with an extremely low ionization potential.

In [115]:
res[res["stage"]=="II"].show_in_notebook()

Certainly no candidates for our neutral lines, although we do pick up the [Cl II] lines at 8578.69 and 9123.6, which we certainly observe, although the shorte one is blended with a He I line.

# Look at the supposed Ca I lines

## The 7890 line

This looks a lot like the [Fe III] lines

In [17]:
wavelength_range = (7890 * u.Angstrom, 0.2 * u.Angstrom)
#wavelength_range = (6500 * u.Angstrom, 6600 * u.Angstrom)
#AtomicLineList.FORM_URL = "https://www.pa.uky.edu/~peter/newpage/"
res = AtomicLineList.query_object(
    wavelength_range=wavelength_range,
    wavelength_type='Air',
    minimal_abundance="6",
    depl_factor="0",
    #element_spectrum='H I',
    #get_query_payload=True,
)

In [18]:
res

<Table length=18>
LAMBDA AIR ANG SPECTRUM  TT  ...   J J     A_ki       LEVEL ENERGY  CM 1   
     str8        str8   str2 ...   str7    str8             str25          
-------------- -------- ---- ... ------- -------- -------------------------
         7608.  Si VII]   E1 ...     1-0     None  1491287.00 -  1504430.00
         7698.    Ar X]   E1 ... 1/2-3/2     None  2555193.00 -  2568152.00
         7788.    Si XI   E1 ...     2-2 1.13E+06  3187600.00 -  3200430.00
         7838.    Si XI   E1 ...     2-3 3.26E+04  3188940.00 -  3201700.00
         7868.    Si XI   E1 ...     2-2 3.61E+05  3188940.00 -  3201650.00
         7873.     N VI   E1 ...     *-* 2.63E+07  4390976.00 -  4403673.70
         7878.    Si XI   E1 ...     2-1 1.59E+05  3200430.00 -  3213120.00
         7888.  Fe III]   E1 ...     2-2     None   167002.00 -   179676.89
        7889.9    N II]   E1 ...     1-2     None   224286.63 -   236957.50
        7889.9 [Ni III]   M1 ...     3-2     None     1360.70 -    14031.60
       7890.07    Ca I]   E1 ...     2-2     None    36575.12 -    49245.80
      7890.154     Ni I   E1 ...     2-1     None    31441.63 -    44112.17
        7890.4     Ar I   E1 ...     3-2     None   113716.55 -   126386.70
         7891.   Al XII   E1 ...     1-2 5.08E+06 15816791.00 - 15829460.00
         7893.      O V   E1 ...     2-3 2.43E+07   818668.00 -   831334.00
         7894.    C III   E1 ...     1-0     None   363613.00 -   376278.10
         7898.   Ca XIX   E1 ...     1-2 5.07E+06 39761380.00 - 39774040.00
        7918.?   Na VII   E1 ...   5/2-* 4.08E+04   1377820.00 -  1390450.?

So it looks like the [Ni III] 7889.9 line is a very good candidate.  It is a nebular forbidden line, so would be expected to be strong. 

In [8]:
AtomicLineList.query_object_async??

Signature:
AtomicLineList.query_object_async(
    wavelength_range=None,
    wavelength_type='',
    wavelength_accuracy=None,
    element_spectrum=None,
    minimal_abundance=None,
    depl_factor=None,
    lower_level_energy_range=None,
    upper_level_energy_range=None,
    nmax=None,
    multiplet=None,
    transitions=None,
    show_fine_structure=None,
    show_auto_ionizing_transitions=None,
    output_columns=('spec', 'type', 'conf', 'term', 'angm', 'prob', 'ener'),
    cache=True,
    get_query_payload=False,
)
Docstring:
Queries Atomic Line List for the given parameters adnd returns the
result as a `~astropy.table.Table`. All parameters are optional.

Parameters
----------
wavelength_range : pair of `astropy.units.Unit` values
    Wavelength range. Can be done in two ways: supply a lower and
    upper limit for the region or, supply the central wavelength and
    the 1 sigma error (68 % confidence value) for that line. If the
    first number is smaller than the second number

In [14]:
aa.Transition.nebular

AtomicTransition<'Neb'>

In [15]:
simple_transitions = [aa.Transition.all, aa.Transition.nebular]

In [16]:
simple_transitions

[AtomicTransition<'All'>, AtomicTransition<'Neb'>]

In [17]:
"Neb" in simple_transitions

False

In [18]:
aa.utils.is_valid_transitions_param("Neb")

False

In [23]:
aa.utils.is_valid_transitions_param(aa.Transition.nebular)

True

In [37]:
from astropy import units as u


In [41]:
u.spectral??

Signature: u.spectral()
Source:   
def spectral():
    """
    Returns a list of equivalence pairs that handle spectral
    wavelength, wave number, frequency, and energy equivalencies.

    Allows conversions between wavelength units, wave number units,
    frequency units, and energy units as they relate to light.

    There are two types of wave number:

        * spectroscopic - :math:`1 / \\lambda` (per meter)
        * angular - :math:`2 \\pi / \\lambda` (radian per meter)

    """
    hc = _si.h.value * _si.c.value
    two_pi = 2.0 * np.pi
    inv_m_spec = si.m ** -1
    inv_m_ang = si.radian / si.m

    return Equivalency([
        (si.m, si.Hz, lambda x: _si.c.value / x),
        (si.m, si.J, lambda x: hc / x),
        (si.Hz, si.J, lambda x: _si.h.value * x, lambda x: x / _si.h.value),
        (si.m, inv_m_spec, lambda x: 1.0 / x),
        (si.Hz, inv_m_spec, lambda x: x / _si.c.value,
         lambda x: _si.c.value * x),
        (si.J, inv_m_spec, lambda x: x / hc, lambda x:

In [43]:
u.Equivalency?

Init signature: u.Equivalency(equiv_list, name='', kwargs=None)
Docstring:     
A container for a units equivalency.

Attributes
----------
name: `str`
    The name of the equivalency.
kwargs: `dict`
    Any positional or keyword arguments used to make the equivalency.
File:           ~/miniconda3/envs/py39/lib/python3.9/site-packages/astropy/units/equivalencies.py
Type:           ABCMeta
Subclasses:     
